CRM/Civicase/Form/CaseCategoryTypeCustomField.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+/**
+ * This class generates form components for custom data for a case category.
+ */
+abstract class CRM_Civicase_Form_CaseCategoryTypeCustomField extends CRM_Core_Form {
+
+  /**
+   * The entity id, used when editing/creating custom data.
+   *
+   * @var int
+   */
+  protected $entityId;
+
+  /**
+   * Entity sub type of the table id.
+   *
+   * @var int
+   */
+  protected $subTypeId;
+
+  /**
+   * Custom group Id.
+   *
+   * @var int
+   */
+  protected $groupId;
+
+  /**
+   * Entity Table.
+   *
+   * @var string
+   */
+  protected $entityTable = 'civicrm_case_type';
+
+  /**
+   * Returns entity type for the case category.
+   *
+   * @return string
+   *   Entity type.
+   */
+  abstract protected function getEntityType();
+
+  /**
+   * Sets the appropriate variables needed by the custom data templates.
+   */
+  public function preProcess() {
+    $this->groupId = CRM_Utils_Request::retrieve('groupId', 'Positive', $this, FALSE, NULL);
+    $this->entityId = CRM_Utils_Request::retrieve('entityId', 'Positive', $this, TRUE);
+
+    $groupTree = CRM_Core_BAO_CustomGroup::getTree(
+      $this->getEntityType(),
+      NULL,
+      $this->entityId,
+      $this->groupId ? $this->groupId : NULL,
+      $this->entityId
+    );
+
+    // Simplified formatted groupTree.
+    $groupTree = CRM_Core_BAO_CustomGroup::formatGroupTree(
+      $groupTree,
+      1,
+      $this
+    );
+
+    foreach ($groupTree as $groupValues) {
```

Comment:
Maybe use it like this to optimise memory usage, what do you think?
```foreach ($groupTree as &$groupValues) {```

Reply:
@ivan-compucorp, not sure what you mean by optimise memory usage? Are you referring to the fact that when we pass it by reference, PHP will not create a copy and use the original variable instead?

I think the code is fine as it is. 
For one, passing by reference is only supposed to be when we intend to modify the original variable in the array itself and not supposed to be a way to optimize memory in itself.
Two, the array size here is small and can never cause issues, most likely the size can never get to 50.